### DISCUSSION TITLE
Linear Interpolation function

### DISCUSSION THREAD

**Discussion by alexis-bvq** [atoti/atoti#506] [General]
Hello, we are facing two different subjects where we need to interpolate a value within a vector.
I can see two different methods for it but the final result is the same.

```py
ladders comparator vector = [0, 10, 20, 30, 40]
ladders values vector  = [5, 23, 52, 34, 57]
```

Then you have a value that you have to place within the comparator ladders in order to get the interpolated value out of the values ladder.

```py
value = 15
```

Then because `10 < 15 < 20` (placing value into ladders vectors)
 => `interpolated_value = ( 23 * (20 - 15) + 52 * (15 - 10) ) / ( 20 - 10 ) = 37.5`

Another example, if `value = 21`:
Then because `20 < 21 < 30` (placing value into ladders vectors) 
 => `interpolated_value = ( 52 * (30 - 21) + 34 * (21 - 20) ) / ( 30 - 20 ) = 50.2`

A typical example would be:
- `ladder comparator vector = ladders of Market Value`
- `ladder values vector = ladders of sensi delta`
- `value = current Market Value`
The pricing engine is anticipating the moves of the market value for one product over the night. Therefore if we take cac40 delta, it would generate:
cac40 -> delta
6000 -> 10
6100 -> 15
6200 -> 25
And according to the current market value of cac40 you can predict delta thanks to the interpolation.

Sometimes the ladders vector and values vectors could be the same.

Sometimes the ladders comparator vector can contain dates, the value is a date too, and therefore you count the date difference in days, months, years.

I think we could create a method in atoti doing that natively, as of now we have to use udaf parsed from java in order to make it work.
It would allow us to produce more use cases for the gallery.

Error handling:
Sometimes a value doesn't have ladders or values vector, sometimes the value is outside of the ladders vector.
This error could return defaults values or `None`.

**Comment by APolyakova** [comment]
Hi! 

This use case is quite common. This is where we needed the calculation:

- intraday sensitivity approximation: risk system computes sensitivities for a bunch of scenarios - spot prices ladder representing market going from -50% down to 50% up - then during the day you observe your actual market prices, and interpolate sensitivities
- rate curve discount factors: you have interest rate curve (implied rate by tenor) and you have projected cash flows with dates. YOu want to compute net present value by multiplying cash flows and their discount factors. To compute the discount factor, you interpolate rate from the interest rate curve. Similarly to rate curves, interpolation is very common for vol surfaces.


Interpolation in python:
- Numpy  https://numpy.org/doc/stable/reference/generated/numpy.interp.html 
- Scipy https://docs.scipy.org/doc/scipy/tutorial/interpolate.html

**Comment by APolyakova** [comment]
We'd like to create blogs on the above topics, so it would be great if we had the function in the API. Thanks!!

**Comment by APolyakova** [comment]
Btw if you could consider implementing a 2d interpolation - we could try to show case some Black Scholes calc, when we combine these with https://docs.atoti.io/latest/lib/atoti.experimental.stats.normal.html#atoti.experimental.stats.normal.pdf 

Not saying it's a good use case for us, because it's a point-level calculation. Perhaps the stress scenarios on the vol surface / rate curve could be an interesting story.

**Comment by rcolle** [comment]
Hi @alexis-bvq and @APolyakova ,
Is this computation always done at fact level or is it sometimes done at an aggregated level (this matters to choose between different designs)? 
Additionally, where do the ladders come from? A store that is joined to the base store?

**Comment by alexis-bvq** [comment]
First of all sorry for the long response time I did not get the notification :/
From my examples it has been done at fact level but it could be done at aggregation level on another example.
Again, in my example the ladders were on a referenced store but they could be on the base store too.
